components/raftstore/src/store/util.rs

```diff
@@ -836,6 +837,163 @@ impl Display for MsgType<'_> {
+    }
+}
+
+#[derive(Debug)]
+struct RegionReadProgressCore {
+    applied_index: u64,
+    safe_ts: u64,
+    // `pending_items` is a *sorted* list of `(apply index, safe ts)` item
+    pending_items: VecDeque<(u64, u64)>,
+    // the max size of `pending_items`
+    cap: usize,
+}
+
+impl RegionReadProgressCore {
+    fn new(applied_index: u64, cap: usize) -> RegionReadProgressCore {
+        RegionReadProgressCore {
+            applied_index,
+            safe_ts: 0,
+            pending_items: VecDeque::with_capacity(cap),
+            cap,
+        }
+    }
+
+    // Return the `safe_ts` if it is updated
+    fn update_applied(&mut self, applied: u64) -> Option<u64> {
+        // The apply index should not decrease
+        assert!(applied >= self.applied_index);
+        self.applied_index = applied;
+        // Consume pending items with `apply_index` less or equal to `self.applied_index`
+        let mut ts_to_update = self.safe_ts;
+        while let Some((idx, ts)) = self.pending_items.pop_front() {
+            if self.applied_index < idx {
+                self.pending_items.push_front((idx, ts));
+                break;
+            }
+            if ts_to_update < ts {
+                ts_to_update = ts;
+            }
+        }
+        if self.safe_ts < ts_to_update {
+            self.safe_ts = ts_to_update;
+            Some(ts_to_update)
+        } else {
+            None
+        }
+    }
+
+    // Return the `safe_ts` if it is updated
+    fn update_safe_ts(&mut self, apply_index: u64, ts: u64) -> Option<u64> {
+        // The peer has enough data, try update `safe_ts` directly
+        if self.applied_index >= apply_index {
+            let mut updated_ts = None;
+            if self.safe_ts < ts {
+                self.safe_ts = ts;
+                updated_ts = Some(ts);
+            }
+            return updated_ts;
+        }
+        // The peer doesn't has enough data, keep the item for future use
+        if let Some(item) = self.pending_items.back_mut() {
+            // Discord the item if it has a smaller `safe_ts`
+            if item.1 >= ts {
+                return None;
+            }
+            // Discord the item if it has a smaller `apply_index`
+            if item.0 >= apply_index {
+                // Instead of adding a new item, try update the last one,
+                // update the last item's `safe_ts` if the incoming item
+                // require a less or equal `apply_index` with a larger `safe_ts`
+                item.1 = ts;
+                return None;
+            }
+        }
+        self.push_back((apply_index, ts));
+        None
+    }
+
+    fn clear(&mut self) {
+        self.pending_items.clear();
+        self.safe_ts = 0;
+    }
+
+    fn push_back(&mut self, item: (u64, u64)) {
+        if self.pending_items.len() >= self.cap {
+            // Randomly remove a item to make room for the incoming one
+            let idx = rand::random::<usize>() % self.cap;
```

Comment:
How about 
```
let mut keep = false;
self.pending_items.retain(|_| {
    keep = !keep;
    keep
});
```
?

Reply:
We only want to remove one item

Reply:
It doesn't matter. Given the capacity is 128, even remove half, there are still 64 safe ts. When the queue is full, it usually means the follower is stuck, so it probably keeps popping for sometime. Shrink it half instead of poping one can reduce the overhead of moving data around again and again.